pkg/cmd/healthprobe/healthprobe.go

```diff
@@ -0,0 +1,316 @@
+			close(shutdownHandler)
+			cancel()
+		}
+	}()
+
+	// enable profiler by default
+	defer serviceability.Profile("web").Stop()
+	serviceability.StartProfiler()
+
+	healthCheck, err := h.newHealthCheck()
+	if err != nil {
+		return err
+	}
+
+	healthCheckProbe := healthCheck.Probe()
+	if err := wait.PollImmediateUntil(h.interval, func() (bool, error) { return healthCheckProbe() }, ctx.Done()); err != nil {
+		// probe is shutting down report last status of existing disruption
+		if err == wait.ErrWaitTimeout && healthCheck.Disruption {
+			healthCheck.lg.Warn("health probe terminated",
+				zap.String("pod", healthCheck.podName),
+				zap.Bool("available", false),
+				zap.Duration("duration", time.Since(healthCheck.DisruptionStart)), // duration
+			)
+		}
+	}
+	healthCheck.lg.Info("health probe is shutting down",
+		zap.String("pod", healthCheck.podName),
+	)
+	return nil
+}
+
+func (h *healthProbeOpts) newHealthCheck() (*healthCheck, error) {
+	if h.enableLogRotation {
+		if err := setupLogRotation(h.logOutputs, h.LogRotationConfigJSON); err != nil {
+			return nil, err
+		}
+	}
+	lg, err := getZapLogger(loglevelToZap(h.logLevel), h.logOutputs, h.enableLogRotation)
+	if err != nil {
+		return nil, err
+	}
+	tlsInfo := transport.TLSInfo{
+		CertFile:      h.clientCertFile,
+		KeyFile:       h.clientKeyFile,
+		TrustedCAFile: h.clientCACertFile,
+	}
+	tr, err := transport.NewTransport(tlsInfo, h.dialTimeout)
+	if err != nil {
+		return nil, err
+	}
+	client := http.Client{
+		Transport: tr,
+	}
+
+	healthCheck := &healthCheck{
+		client:            client,
+		endpoint:          h.Endpoint,
+		lg:                lg,
+		interval:          h.interval,
+		podName:           h.podName,
+		staticPodVersion:  h.staticPodVersion,
+		enableLogRotation: h.enableLogRotation,
+	}
+	return healthCheck, nil
+}
+
+func (h *healthCheck) Probe() HealthCheckFunc {
+	h.lg.Warn("starting health probe",
+		zap.String("pod", h.podName),
+	)
+	return func() (bool, error) {
+		isHealthy, _ := checkHealth(h.client, h.endpoint, h.podName, h.lg)
+		// track new service disruption
+		if !isHealthy && !h.Disruption {
+			h.Disruption = true
+			h.DisruptionStart = time.Now()
+			h.lg.Warn(
+				"service disruption detected",
+				zap.String("pod", h.podName),
+				zap.Int("version", h.staticPodVersion),
+				zap.Bool("available", false),
+				zap.Duration("duration", 0*time.Second),
+			)
+		}
+		// report previous disruption duration
+		if isHealthy && h.Disruption {
+			// clear disruption
+			h.lg.Warn(
+				"service disruption restored",
+				zap.String("pod", h.podName),
+				zap.Int("version", h.staticPodVersion),
+				zap.Bool("available", true),
+				zap.Duration("duration", time.Since(h.DisruptionStart)),
+			)
+			h.Disruption = false
+			h.DisruptionStart = time.Time{}
+		}
+		return false, nil
+	}
+}
+
+func checkHealth(client http.Client, endpoint string, podName string, lg *zap.Logger) (bool, error) {
+	resp, err := client.Get(endpoint + "/health")
```

Comment:
Can we make this a quorum read of some key like the kube-apiserver would do?  namespaces/default perhaps?

The individual health seems like a separate (and valuable) check.

Reply:
yes, makes sense.

Reply:
done